src/ucp/core/ucp_am.c

```diff
@@ -1156,30 +1189,39 @@ ucp_am_invoke_cb(ucp_worker_h worker, uint16_t am_id, void *user_hdr,
     return am_cb->cb_old(am_cb->context, data, data_length, reply_ep, flags);
 }
 
-static UCS_F_ALWAYS_INLINE ucs_status_t
-ucp_am_handler_common(ucp_worker_h worker, ucp_am_hdr_t *am_hdr, size_t hdr_size,
-                      size_t total_length, ucp_ep_h reply_ep, unsigned am_flags,
-                      uint64_t recv_flags)
+static UCS_F_ALWAYS_INLINE ucs_status_t ucp_am_handler_common(
+        ucp_worker_h worker, ucp_am_hdr_t *am_hdr, size_t total_length,
+        ucp_ep_h reply_ep, unsigned am_flags, uint64_t recv_flags)
 {
     ucp_recv_desc_t *desc    = NULL;
     uint16_t am_id           = am_hdr->am_id;
     uint32_t user_hdr_size   = am_hdr->header_length;
-    ucp_am_entry_t *am_cb    = &ucs_array_elem(&worker->am, am_id);
-    void *data               = UCS_PTR_BYTE_OFFSET(am_hdr, hdr_size);
+    ucp_am_entry_t *am_cb    = &ucs_array_elem(&worker->am.cbs, am_id);
+    void *data               = am_hdr + 1;
     size_t data_length       = total_length -
-                               (hdr_size + am_hdr->header_length);
+                               (sizeof(*am_hdr) + am_hdr->header_length);
     void *user_hdr           = UCS_PTR_BYTE_OFFSET(data, data_length);
     ucs_status_t desc_status = UCS_OK;
     ucs_status_t status;
 
-    ucs_assert(total_length >= am_hdr->header_length + hdr_size);
+    ucs_assert(total_length >= am_hdr->header_length + sizeof(*am_hdr));
 
     /* Initialize desc in advance, so the user could invoke ucp_am_recv_data_nbx
      * from the AM callback directly. The only exception is inline data when
      * AM callback is registered without UCP_AM_FLAG_PERSISTENT_DATA flag.
      */
     if ((am_flags & UCT_CB_PARAM_FLAG_DESC) ||
         (am_cb->flags & UCP_AM_FLAG_PERSISTENT_DATA)) {
+
+        /* UCT may not support AM data alignment. If unaligned data ptr is
+         * provided in UCT descriptor, allocate new aligned data buffer from UCP
+         * AM mpool instead of using UCT descriptor directly.
+         */
+        if (ucs_unlikely(((uintptr_t)data % worker->am.alignment) &&
+                         (am_flags & UCT_CB_PARAM_FLAG_DESC))) {
+            am_flags &= ~UCT_CB_PARAM_FLAG_DESC;
+        }
```

Comment:
1. maybe need to configure somehow if this "alignment fix" should be done by ucp?
2. IMO no need to check `(am_flags & UCT_CB_PARAM_FLAG_DESC)`

Reply:
1. if it is not needed `data % worker->am.alignmen` should be 0, because `worker->am.alignmen` would be 1
2. 👍

Reply:
you mean `worker->am.alignment` is mandatory and not advisory?

Reply:
it's 1 by default (if no specific alignment requested)